requests_openapi_client/schemas.py

```diff
@@ -106,10 +119,46 @@ def type_for_schema(schema, full_spec, realized_types={}):
 
     return TYPE_EQUIVALENTS.get(schema["type"], None)
 
+class TaggedUnion:
+    discriminator_field: str
+    discriminator_mapping: typing.Mapping[typing.Any, typing.Any]
+
+    def __init__(self, type_options, discriminator, full_spec, realized_types={}):
+        self.discriminator_field = discriminator.get("propertyName", None)
+        mapping = discriminator.get("mapping", None)
+
+        if not self.discriminator_field or not mapping:
+            raise Exception("discriminator property name and mapping are required")
+
+        types_by_ref = {type_option.get("$ref", None): type_for_schema(type_option, full_spec, realized_types) for type_option in type_options}
+        self.discriminator_mapping = {name: types_by_ref[ref] for name, ref in mapping.items()}
+
+    @property
+    def allowed_types(self):
+        return list(self.discriminator_mapping.values())
+
+    @property
+    def type_for_annotation(self):
+        return typing.Union.__getitem__(tuple(self.allowed_types))
+
+    def deserialize(self, data):
+        type_value = data.get(self.discriminator_field, None)
+        if not type_value:
+            raise Exception("discriminator mapping field not found on union object")
+        if type_value not in self.discriminator_mapping:
+            raise Exception(f"discriminator value {type_value} not in union")
+        return deserialize_as(data, self.discriminator_mapping[type_value])
+
 def update_field_type(model, field_name, new_type):
+    model.__dataclass_fields__[field_name].type = new_type
+
+    if hasattr(new_type, "type_for_annotation"):
```

Comment:
This I don't fully grok how it works - I understand the goal described in the comment, but not so much the mechanics of how it functions.

Reply:
@bvacherot-sofar yeah, the gist is that we (maybe mostly I, not sure if anyone else cares :sweat_smile: ) want to be able to play with these endpoints from the Python REPL, and be able to do, like, `help(client.voyages.get_voyage)` or whatever and see what arguments it expects. We do this dance where we're gradually building up these functions, methods, etc., and at the time where we build each one, it's possible that the types we want to use as the expected types here (that'll eventually get displayed in the REPL) may not exist yet. So initially we build out all the types with placeholders, then we circle back once they're all built and rewrite all the annotations so they point to the now-instantiated types.

We _also_ use the types of each field for a different purpose, which is figuring out how to serialize and deserialize each value (we need to know which kind of object to instantiate given a chunk of JSON in an API response, for example). Up until now, we've been able to use the same type for each of these purposes: for deserialization, and also for annotation. Unfortunately, for this new `oneOf` usecase, that was no longer possible: what we want for annotation purposes is for the type to be `typing.Union[X, Y, Z]` where X, Y, and Z are the allowed types in a given field. But we can't use that for deserialization. For one thing, it's not possible to instantiate a `typing.Union`, and for another thing, we need an extra bit of information at deserialization time: the `discriminator` info that lets us tell which of X, Y, or Z we should use for deserializing a given value. So: there's now an internal class that effectively captures both the allowed types and the discriminator info that we use for serialization/deserialization, and it supplies, via a property, what type should be displayed in the REPL (the appropriate `typing.Union`). So this mechanism checks to see if the type should be annotated as some other type, and if so, uses that.